tempodb/encoding/vparquet4/combiner.go

```diff
@@ -0,0 +1,191 @@
+package vparquet4
+
+import (
+	"bytes"
+	"sort"
+
+	"github.com/grafana/tempo/pkg/util"
+)
+
+func combineTraces(traces ...*Trace) *Trace {
+	if len(traces) == 1 {
+		return traces[0]
+	}
+
+	c := NewCombiner()
+	for i := 0; i < len(traces); i++ {
+		c.ConsumeWithFinal(traces[i], i == len(traces)-1)
+	}
+	res, _, _ := c.Result() // for now ignore the connected return. see comment in walBlock.Iterator()
+	return res
+}
+
+// Combiner combines multiple partial traces into one, deduping spans based on
+// ID and kind.  Note that it is destructive. There are design decisions for
+// efficiency:
+// * Only scan/hash the spans for each input once, which is reused across calls.
+// * Only sort the final result once and if needed.
+// * Don't scan/hash the spans for the last input (final=true).
+type Combiner struct {
+	result   *Trace
+	spans    map[uint64]struct{}
+	combined bool
+}
+
+func NewCombiner() *Combiner {
+	return &Combiner{}
+}
+
+// Consume the given trace and destructively combines its contents.
+func (c *Combiner) Consume(tr *Trace) (spanCount int) {
+	return c.ConsumeWithFinal(tr, false)
+}
+
+// ConsumeWithFinal consumes the trace, but allows for performance savings when
+// it is known that this is the last expected input trace.
+func (c *Combiner) ConsumeWithFinal(tr *Trace, final bool) (spanCount int) {
+	if tr == nil {
+		return
+	}
+
+	// First call?
+	if c.result == nil {
+		c.result = tr
+
+		// Pre-alloc map with input size. This saves having to grow the
+		// map from the small starting size.
+		n := 0
+		for _, b := range c.result.ResourceSpans {
+			for _, ils := range b.ScopeSpans {
+				n += len(ils.Spans)
+			}
+		}
+		c.spans = make(map[uint64]struct{}, n)
+
+		for _, b := range c.result.ResourceSpans {
+			for _, ils := range b.ScopeSpans {
+				for _, s := range ils.Spans {
+					c.spans[util.SpanIDAndKindToToken(s.SpanID, s.Kind)] = struct{}{}
+				}
+			}
+		}
+		return
+	}
+
+	// coalesce root level information
+	if tr.EndTimeUnixNano > c.result.EndTimeUnixNano {
+		c.result.EndTimeUnixNano = tr.EndTimeUnixNano
+	}
+	if tr.StartTimeUnixNano < c.result.StartTimeUnixNano || c.result.StartTimeUnixNano == 0 {
+		c.result.StartTimeUnixNano = tr.StartTimeUnixNano
+	}
+	if c.result.RootServiceName == "" {
+		c.result.RootServiceName = tr.RootServiceName
+	}
+	if c.result.RootSpanName == "" {
+		c.result.RootSpanName = tr.RootSpanName
+	}
+	c.result.DurationNano = c.result.EndTimeUnixNano - c.result.StartTimeUnixNano
+
+	// Merge service stats
+	for service, incomingStats := range tr.ServiceStats {
+		combinedStats := c.result.ServiceStats[service]
+		combinedStats.SpanCount += incomingStats.SpanCount
```

Comment:
i don't think we can just sum here. i think we have to rescan the trace and recalculate similar to nested set.

Reply:
Should we apply the same logic as in https://github.com/stoewer/tempo/blob/add-vparquet4-encoding/tempodb/encoding/vparquet4/schema.go#L560-L575 with the Trace in `c.result` (and moving this to the end of the function, after the loop where the spans are copied)?

Reply:
yes. notice the function `assignNestedSetModelBounds` it's called in `traceToParquet` (when we first create a trace) and in the combiner.

Reply:
It could also be a good idea to rename the function to `assignNestedSetModelBoundsAndServiceStats` and move the logic there. The downside is, that it adds complexity to an already complex function. But it also has two advantages:
* It makes sure that service stats are always (re-) calculated when needed
* `assignNestedSetModelBounds` already iterates through the whole trace. Moving the logic there, eliminates a second traversal of the trace

@joe-elliott what do you think?

Reply:
> @joe-elliott what do you think?

+1 do this. i could see more "recalculate after combining" summary columns being added. co-locating them in one spot for performance feels good to me.

it is a relatively easy to function to test. some clear tests should help with the complexity concerns.

Reply:
:+1:

I moved the logic to `assignNestedSetModelBoundsAndServiceStats()` and added a few new testcases: https://github.com/stoewer/tempo/pull/3